usr/sbin/p11sak/p11tool.c

```diff
@@ -0,0 +1,2377 @@
+            (*longopts_count)++;
+    }
+}
+
+static CK_RV p11tool_build_opts(const struct p11tool_opt *opts,
+                                char *optstring,
+                                struct option *longopts)
+{
+    const struct p11tool_opt *opt;
+    unsigned int opts_idx, long_idx;
+
+    opts_idx = strlen(optstring);
+
+    for (long_idx = 0; longopts[long_idx].name != NULL; long_idx++)
+        ;
+
+    for (opt = opts; opt->short_opt != 0 || opt->long_opt != NULL; opt++) {
+        if (opt->short_opt != 0) {
+            optstring[opts_idx++] = opt->short_opt;
+            if (opt->arg.type != ARG_TYPE_PLAIN) {
+                optstring[opts_idx++] = ':';
+                if (!opt->arg.required)
+                    optstring[opts_idx++] = ':';
+            }
+        }
+
+        if (opt->long_opt != NULL) {
+            longopts[long_idx].name = opt->long_opt;
+            longopts[long_idx].has_arg = opt->arg.type != ARG_TYPE_PLAIN ?
+                              (opt->arg.required ?
+                                      required_argument : optional_argument ) :
+                              no_argument;
+            longopts[long_idx].flag = NULL;
+            longopts[long_idx].val = opt->short_opt != 0 ?
+                                        opt->short_opt : opt->long_opt_val;
+            long_idx++;
+        }
+    }
+
+    return CKR_OK;
+}
+
+static CK_RV p11tool_build_cmd_opts(const struct p11tool_opt *cmd_opts,
+                                    const struct p11tool_opt *generic_opts,
+                                    char **optstring, struct option **longopts)
+{
+    unsigned int optstring_len = 0, longopts_count = 0;
+    CK_RV rc;
+
+    p11tool_count_opts(generic_opts, &optstring_len, &longopts_count);
+    if (cmd_opts != NULL)
+        p11tool_count_opts(cmd_opts, &optstring_len, &longopts_count);
+
+    *optstring = calloc(1 + optstring_len + 1, 1);
+    *longopts = calloc(longopts_count + 1, sizeof(struct option));
+    if (*optstring == NULL || *longopts == NULL) {
+        rc = CKR_HOST_MEMORY;
+        goto error;
+    }
+
+    (*optstring)[0] = ':'; /* Let getopt return ':' on missing argument */
+
+    rc = p11tool_build_opts(generic_opts, *optstring, *longopts);
+    if (rc != CKR_OK)
+        goto error;
+
+    if (cmd_opts != NULL) {
+        rc = p11tool_build_opts(cmd_opts, *optstring, *longopts);
+        if (rc != CKR_OK)
+            goto error;
+    }
+
+    return CKR_OK;
+
+error:
+    if (*optstring != NULL)
+        free(*optstring);
+    *optstring = NULL;
+
+    if (*longopts != NULL)
+        free(*longopts);
+    *longopts = NULL;
+
+    return rc;
+}
+
+static CK_RV p11tool_process_plain_argument(const struct p11tool_arg *arg)
+{
+    *arg->value.plain = true;
+
+    return CKR_OK;
+}
+
+static CK_RV p11tool_process_string_argument(const struct p11tool_arg *arg,
+                                             char *val)
+{
+    *arg->value.string = val;
```

Comment:
arg is const, but modified by assigning val? Same above. And you intentionally use *arg-> ? I would expect arg->

Reply:
It is NOT modifying `arg`. It is only modifying where the constant pointer `arg->value.string` points to. 
Thus the `*arg->value.string`....